src/python/dxpy/bindings/dxfile_functions.py

```diff
@@ -537,33 +500,89 @@ def report_progress(handler, num_bytes):
+        # already
+        creation_kwargs = kwargs.copy()
+        if 'name' not in kwargs:
+            if filename is not None:
+                creation_kwargs['name'] = os.path.basename(filename)
+            else:
+                # Try to get filename from file-like object
+                try:
+                    local_file_name = file.name
+                except AttributeError:
+                    pass
+                else:
+                    creation_kwargs['name'] = os.path.basename(local_file_name)
 
-    while True:
-        buf = read(handler._write_bufsize)
-        offset += len(buf)
+        # Use 'a' mode because we will be responsible for closing the file
+        # ourselves later (if requested).
+        return new_dxfile(mode='a', media_type=media_type, write_buffer_size=write_buffer_size,
+                             expected_file_size=file_size, file_is_mmapd=file_is_mmapd, **creation_kwargs)
 
-        if len(buf) == 0:
-            break
+    retries = 0
+    max_retries = 2
+    file_is_mmapd = hasattr(fd, "fileno")
 
-        handler.write(buf,
-                      report_progress_fn=report_progress if show_progress else None,
-                      multithread=multithread,
-                      **remaining_kwargs)
+    if write_buffer_size is None:
+        write_buffer_size=dxfile.DEFAULT_BUFFER_SIZE
 
-    if filename is not None:
-        fd.close()
+    while retries <= max_retries:
+        retries += 1
 
-    handler.flush(report_progress_fn=report_progress if show_progress else None, **remaining_kwargs)
+        if use_existing_dxfile:
+            handler = use_existing_dxfile
+        else:
+            handler = get_new_handler(filename)
 
-    if show_progress:
-        sys.stderr.write("\n")
-        sys.stderr.flush()
+        # For subsequent API calls, don't supply the dataobject metadata
+        # parameters that are only needed at creation time.
+        _, remaining_kwargs = dxpy.DXDataObject._get_creation_params(kwargs)
 
+        num_ticks = 60
+        offset = 0
+
+        handler._ensure_write_bufsize(**remaining_kwargs)
+
+        handler._num_bytes_transmitted = 0
+
+        if show_progress:
+            report_progress(handler, 0)
+
+        while True:
+            buf = read(handler._write_bufsize)
+            offset += len(buf)
+
+            if len(buf) == 0:
+                break
+
+            handler.write(buf,
+                          report_progress_fn=report_progress if show_progress else None,
+                          multithread=multithread,
+                          **remaining_kwargs)
+
+        handler.flush(report_progress_fn=report_progress if show_progress else None, **remaining_kwargs)
+
+        if show_progress:
+            sys.stderr.write("\n")
+            sys.stderr.flush()
+        try:
+            handler.wait_until_parts_uploaded()
+            if filename is not None:
+                fd.close()
+
+            break
+        except DXError:
```

Comment:
Whats happening here?

In case a part did not finish successfully - it is not in completed state, we will try to reupload the file. I chose 2 retries.